### ISSUE TITLE
Passing null parameters to x:Bind function binding

### ISSUE THREAD

**Issue by vsfeedback** [microsoft/microsoft-ui-xaml#2166]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/961183/passing-null-parameters-to-xbind-function-binding.html)._

---
<p>I'm trying to bind an object using x:Bind function binding. See <a href="https://docs.microsoft.com/en-us/windows/uwp/data-binding/function-bindings">https://docs.microsoft.com/en-us/windows/uwp/data-binding/function-bindings</a>.</p> <p>This works fine as long as the parameter has a value. But if the parameter is null, the function is not called and the binding uses FallbackValue instead (if it is defined).</p> <pre>&lt;TextBlock Text=&quot;{x:Bind local:MainViewModel.FormatWidget(ViewModel.SelectedItem), Mode=OneWay, FallbackValue=MyFallbackValue}&quot; /&gt; </pre><p>Interestingly, if the <strong>initial </strong>value is null, x:Bind <strong>does </strong>call the function with the null parameter, but on any subsequent changes when the parameter becomes null, the function is <strong>not </strong>called. It is for this one reason that I would call this a bug rather than a feature request. Surely this is not the intended behavior.</p> <p>Please fix x:Bind to accept null parameters.</p>

---
#### Visual Studio Feedback System on 3/24/2020, 10:27 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by Sergio0694** [comment]
+1 to this, there seem to be no proper way to have fine-tuned control for `null` values with `x:Bind` to function.
And the behavior itself also seems inconsistent as OP mentioned, with the function being invoked **only the first time** with just `null` as argument, and then further invocations just skipped, with `FallbackValue` being used. I think if `FallbackValue` is to be supported in this case, it should always be used, even for the initial binding update.

In my case, I was setting up some samples for the MVVM Toolkit library (see https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/3230), and came up with this:

```xml
<Run
    xmlns:ex="using:Microsoft.Toolkit.Extensions"
    Text="{x:Bind ex:TaskExtensions.GetResultOrDefault(ViewModel.DownloadTextCommand.ExecutionTask), Mode=OneWay}"/>
```

That `DownloadTextCommand` is an `IAsyncRelayCommand` instance (from the same library). That `ExecutionTask` is a `Task<T>` that is **never `null`**. The issue is that the code above just throws a `NullReferenceException`, as the `x:Bind` code always calls that function that's bound to, no matter the input value. You can see this in the codegen:

![image](https://user-images.githubusercontent.com/10199417/89329589-500e4000-d68f-11ea-82e7-3a45c5b1d463.png)

The binding is only checking the _parent_ for `null` values, but not the leaf property being accessed.
I tried with `FallbackValue=''`, but that didn't solve the issue - still got a NRE on initialization.

Any thoughts on this? 🤔
Thanks!

**Comment by JeanRoca** [comment]
Thanks for filing this bug. We will look at fixing this post WinUI 3.0.

**Comment by dojo90** [comment]
1+

**Comment by hawkerm** [comment]
Also Related to #1904, I think (may be same root cause?)

Just hit this again and took a few minutes and digging back into the debugger to realize it just wasn't working as expected. This seems like a pretty weird case. Would be nice to get a fix for both UWP and WinUI 3.

Digging into the generated x:Bind code, I can see during initialization it calls a different method that has an aggregate of function calls, here the null check is on something else. However, when the property change gets called, there's a different null check in a different place which causes it not to cascade down to the function call.

It's pretty easy to spot this difference when following up the tree from the generated function wrapper just a couple of levels.

**Comment by kinwong** [comment]
+1

**Comment by kinwong** [comment]
1+

**Comment by heartacker** [comment]
this is BUG.
MS Team should take it high priority

**Comment by Empressia** [comment]
I am trying to bind multiple variables that can be null in WinUI3,
but this is not possible due to the lack of IMultiValueConverter and this bug.

**Comment by myblindy** [comment]
Exactly, `IMultiValueConverter` died for us to have functions, but they don't work with `null` at all.

**Comment by brooke-tilley** [comment]
Bumping as this is still a serious issue in WinU 3 that makes function bindings significantly less viable as a replacement for Converters

**Comment by greatoceansoftware** [comment]
Same problem here. Would really like this to work. Calling this.Bindings.Update on the page code behind seems to force the OneWay binding to the function to work with null argument, but really shouldn't need this workaround.

**Comment by sjb-sjb** [comment]
So, this bug seems to mean that throughout my entire codebase, wherever  I bind to an object that could be null, I must either specify a Converter that handles null or I must (very redundantly) specify FallbackValue={x:Null} to handle the case where the source object is set to null. Otherwise my x:Bind will not update when the source becomes null.  Certainly a bug that needs to be fixed !!

**Comment by greatoceansoftware** [comment]
Any movement here? Would really like this to be resolved. Thanks.

**Comment by sjb-sjb** [comment]
Yeah really! Was just bitten by this bug AGAIN.
Having to specify FallbackValue={x:Null} everywhere is a complete pain in the neck. 
@michael-hawker what is the state of this -- "moved this to Ready" but ...
